### ISSUE TITLE
Apply proper ARIA attributes to the data grid filter

### ISSUE THREAD

**Issue by bashibozuk** [vmware/clarity#3985]
## Summary

The filter control does not convey an expanded/collapsed state to inform screen reader users that the control expands content when activated.

In addition, the Filter pop-up that appears after activating the Filter control does not have an appropriate name or role.

Screen reader users will not be aware that expandable content is present, or be able to recognize the state of the expandable content after interacting with the control. In addition, the lack of an informative name and role on the Filter pop-up makes it difficult for screen reader users to identify where pop-up content begins/ends.

* What is the change?
- Apply an aria-expanded attribute with a value of true when content is expanded and false when content is collapsed.
- Apply a role="region" to the pop-up container and provide a meaningful name for the container using aria-label.
* Why should it go in Clarity?
Accessibility improvement
* Does this change impact existing behaviors? If so how?
No
* If this change introduces a new behavior, is this behavior accessible?
Yes

## Use case

As a screen reader user I want to be notified that the column on which heading I am on is filterable

## Examples

N/A
## Workarounds

N/A

**Comment by gnomeontherun** [comment]
@chlane can you review this and advise?

**Comment by chlane** [comment]
This issue was reported in issue 168 on http://cx-guidelines.vmware.com/clarityTesting/ and in https://github.com/vmware/clarity/issues/3349. See the original report below. My understanding is that the filter itself is provided by the Clarity Consumer and was out of scope for testing. But I agree that the filter itself needs the fixes that @bashibozuk recommends. Additionally, the color swatches in our example are not keyboard accessible and appear to have low contrast.  There were also focus management issue mentioned in issue #3 at http://cx-guidelines.vmware.com/clarityTesting/ and in https://github.com/vmware/clarity/issues/3370. 

Test URL: https://v1.clarity.design/datagrid/custom-filtering
Clarity Version: v2.0.0-rc.1
Angular Version: 8.0.0-rc.1
OS: any
Browser: any
Screen Reader: VoiceOver, NVDA 2018.41

WCAG 2.1 Guideline:
4.1.2 Name, Role, Value
https://www.w3.org/TR/WCAG21/#name-role-value
(Level A)


Description:
The datagrid-filter-toggle button lacks an accessible name and state as seen below.

<button class="datagrid-filter-toggle" type="button"><clr-icon class="is-solid" shape="filter-grid" role="none"><svg version="1.1" class="has-solid " viewBox="0 0 36 36" preserveAspectRatio="xMidYMid meet" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" focusable="false" role="img"><path class="clr-i-outline clr-i-outline-path-1" d="M15,25.86l2,1V20.27a1,1,0,0,0-.29-.7L10.23,13H25.79l-6.47,6.57a1,1,0,0,0-.29.7L19,28l2,1V20.68L27.58,14A1.46,1.46,0,0,0,28,13V12a1,1,0,0,0-1-1H9a1,1,0,0,0-1,1v1a1.46,1.46,0,0,0,.42,1L15,20.68Z"></path>
                <path class="clr-i-solid clr-i-solid-path-1" d="M8,11v1.12a.5.5,0,0,0,.15.35l7.28,7.36a.5.5,0,0,1,.15.35v6.89a.5.5,0,0,0,.28.45l3.95,1.41a.5.5,0,0,0,.72-.45l0-8.39a.54.54,0,0,1,.18-.35l7.12-7.25a.5.5,0,0,0,.15-.35V11Z"></path></svg></clr-icon></button>



Steps to Reproduce:
- Using a screen reader, go to https://v1.clarity.design/datagrid/custom-filtering
- Read the datagrid-filter-toggle button
- Observe that you do not hear the purpose or state of the button


Recommendation:
Use aria-label to label this button with text that describes the purpose of the button and use aria-expanded to indicate its state

**Comment by gnomeontherun** [comment]
This may already be resolved in latest versions of Clarity with the new popover implementation and only found on v1. @hippee-lee will confirm.

**Comment by lf-collins** [comment]
I'm using VoiceOver to test the datagrid filter, and I found that screen reading always says "edit text black" when I focus on the filter input box. I think we should also add the aria-label attribute to all filter inputs, and each input should preferably have a specific aria-label value, for example, the filter input of “Name" column is “Filter name column”, and it is “Filter age column.” for “Age” column. This way, blind people will always know which column they are filtering. If we cannot add a specific aria-label value, we should at least add the aria-label attribute to filter input.

**Comment by hippee-lee** [comment]
@chlane - the comment above is asking for something different but related (in where it is in the templates and codebase) from the reported issue. 

Can you weigh in on if the commented request is something we want/need to address? If it is something to add, should it be in the pull request that is currently in review for this or should I open a separate issue for it?

**Comment by chlane** [comment]
I agree with @lf-collins. While these inputs are visibly labelled by the combination of the shape="filter-grid" icon and the column header text, they are not programatically labelled. Therefore, screen reader users won't hear the purpose of these inputs when they read them. The programmatic labels must be unique.    

aria-label is one approach. If we use this approach we should follow @lf-collins's guidance on the aria-label values.

Another approach is to use aria-label and aria-labelledby which would allow us to use the consumer injected content along with shape="filter-grid" icon to compose the programmatic label.

See the HTML below.


```
<clr-icon id="filterShape1" shape="filter-grid" role="img" aria-label="filter column" />

<span id="column1Name">Favorite color</span>

<input aria-labelledby="filterShape1 column1Name" />
```

**Comment by hippee-lee** [comment]
Thaks @clane. This will have to be a separate issue. I'll open one up with these notes.

**Comment by hippee-lee** [comment]
> I'm using VoiceOver to test the datagrid filter, and I found that screen reading always says "edit text black" when I focus on the filter input box. I think we should also add the aria-label attribute to all filter inputs, and each input should preferably have a specific aria-label value, for example, the filter input of “Name" column is “Filter name column”, and it is “Filter age column.” for “Age” column. This way, blind people will always know which column they are filtering. If we cannot add a specific aria-label value, we should at least add the aria-label attribute to filter input.

@lf-collins - can you confirm where you are testing this behavior at and which filter you are referring to. The website demo It's hard to be sure this isn't something that should already work with the a11y improvements we made to Clarity already. 

It would really help if you can provide a standalone stackblitz that focuses only on this issue: https://stackblitz.com/@clr-team

**Comment by lf-collins** [comment]
Hi @hippee-lee and  @chlane , thanks for your quick reply!

A good example is the Clarity document, see https://clarity.design/documentation/datagrid/full.
When focusing on the Name column filter button, filter input box, and close button, VoiceOver says "button", "edit text blank", and "Close button", see screenshots:
<img width="1435" alt="Screen Shot 2020-01-16 at 10 33 55 AM" src="https://user-images.githubusercontent.com/26111301/72488448-f443ad00-384b-11ea-9312-4bcd9f87a296.png">
<img width="1440" alt="Screen Shot 2020-01-16 at 10 34 14 AM" src="https://user-images.githubusercontent.com/26111301/72488457-fa398e00-384b-11ea-96ed-28a4ae76c7a8.png">
<img width="1137" alt="Screen Shot 2020-01-16 at 10 38 51 AM" src="https://user-images.githubusercontent.com/26111301/72488705-b72bea80-384c-11ea-8e74-395e6bf967d6.png">
The behavior is the same when filtering other columns. In this case, the blind do not know which column they are dealing with.

Expected behavior is that we specify column name for each action, such as "Open name filter(I am not sure, this button is also used to close the filter box)", "Filter name", and "Close name filter".

When I use the Tab key to move between column headings, VoiceOver doesn't tell me which column I'm on
Another question is not part of this issue but related. When I move between column header using Tab key, VoiceOver doesn't tell me which column I am on. If we can tell user the column name when they enter it, it would also help them to know the which column they are filtering or even sorting.

**Comment by lf-collins** [comment]
Hi @hippee-lee and @chlane , thanks for your quick reply!

A good example is the Clarity document, see https://clarity.design/documentation/datagrid/full.
When focusing on the Name column filter button, input box, and close button, VoiceOver says "button", "edit text blank", and "Close button", see screenshots:
<img width="1435" alt="Screen Shot 2020-01-16 at 10 33 55 AM" src="https://user-images.githubusercontent.com/26111301/72489664-ac268980-384f-11ea-8524-e8acac38ec3a.png">
<img width="1440" alt="Screen Shot 2020-01-16 at 10 34 14 AM" src="https://user-images.githubusercontent.com/26111301/72489665-ac268980-384f-11ea-96c0-d0ddb28027d6.png">
<img width="1137" alt="Screen Shot 2020-01-16 at 10 38 51 AM" src="https://user-images.githubusercontent.com/26111301/72489666-acbf2000-384f-11ea-9a51-5dab250881a6.png">
The behavior is the same when filtering other columns. In this case, the blind do not know which column they are dealing with.

Expected behavior is that we specify column name for each action, such as "Open name filter(I am not sure, this button is also used to close the filter box)", "Filter name", and "Close name filter".

Another question is not part of this issue but related. When I use the Tab key to move between column headings, VoiceOver doesn't tell me which column I'm on. If we can tell users the column name when they enter it, it will also help them know which column they are filtering or even sorting.

**Comment by clane** [comment]
Thanks @lf-collins! 

Using the Tab key to focus the grid headers is not the ARIA specified way to interact with them. I have recommend to provide arrow key support for moving focus (or activedescendant) around the grid in https://github.com/vmware/clarity/issues/3349.
